README.rst
Exile: Python YubiKey AWS signature library
===========================================

**Exile** stores your AWS access key on your YubiKey device and uses it to sign your AWS API requests, protecting you
against credential theft.

Installation
------------
::

    pip install exile

On Linux, install `pcsc-lite <https://salsa.debian.org/rousseau/PCSC>`_
(``apt install pcscd``, ``yum install pcsc-lite``).

Exile requires Python 3.6+.

Synopsis
--------

.. code-block:: python

    import boto3, botocore.auth
    from exile import YKOATH, botocore_signers

    def write_active_aws_key_to_yubikey():
        credentials = boto3.Session().get_credentials()

        key_name = "exile-{}-SigV4".format(credentials.access_key)
        secret = b"AWS4" + credentials.secret_key.encode()
        print("Writing YubiKey OATH SigV4 credential", key_name, "for", credentials.access_key)
        YKOATH().put(key_name, secret, algorithm=YKOATH.Algorithm.SHA256)

        key_name = "exile-{}-HmacV1".format(credentials.access_key)
        secret = credentials.secret_key.encode()
        print("Writing YubiKey OATH HmacV1 credential", key_name, "for", credentials.access_key)
        YKOATH().put(key_name, secret, algorithm=YKOATH.Algorithm.SHA1)

    write_active_aws_key_to_yubikey()
    botocore_signers.install()

    print("Using YubiKey credential to perform AWS call")
    print(boto3.client("sts").get_caller_identity())

    print("Using YubiKey credential to presign an S3 URL")
    print(boto3.client("s3").generate_presigned_url(ClientMethod="get_object", Params={"Bucket": "foo", "Key": "bar"}))

Storing the secret key on a YubiKey instead of in the home directory (``~/.aws/credentials``) protects it in case the
host computer or its filesystem is compromised. The YubiKey acts as an `HSM
<https://en.wikipedia.org/wiki/Hardware_security_module>`_, and can optionally be further configured to require user
interaction (pressing a button on the key) to sign the request::

    YKOATH().put(key_name, secret, algorithm=YKOATH.Algorithm.SHA256, require_touch=True)

TOTP
----

Because exile uses the `YubiKey OATH <https://developers.yubico.com/OATH/>`_ protocol, you can also use it to store
`TOTP <https://en.wikipedia.org/wiki/Time-based_One-time_Password_algorithm>`_
`2FA <https://en.wikipedia.org/wiki/Multi-factor_authentication>`_ tokens, generate and verify codes::

    from exile import TOTP
    TOTP().save("google", "JBSWY3DPEHPK3PXP")  # Or TOTP.save_otpauth_uri("otpauth://...")
    TOTP().get("google")  # Returns a standard 6-digit TOTP code as a string
    TOTP().verify("260153", label="google", at=datetime.datetime.fromtimestamp(1297553958))

Authors
-------
* Andrey Kislyuk

Links
-----
* `Project home page (GitHub) <https://github.com/pyauth/exile>`_
* `Documentation (Read the Docs) <https://exile.readthedocs.io/en/latest/>`_
* `Package distribution (PyPI) <https://pypi.python.org/pypi/exile>`_
* `Change log <https://github.com/pyauth/exile/blob/master/Changes.rst>`_

Bugs
----
Please report bugs, issues, feature requests, etc. on `GitHub <https://github.com/pyauth/exile/issues>`_.

License
-------
Licensed under the terms of the `Apache License, Version 2.0 <http://www.apache.org/licenses/LICENSE-2.0>`_.

.. image:: https://img.shields.io/travis/com/pyauth/exile.svg
        :target: https://travis-ci.com/pyauth/exile
.. image:: https://codecov.io/github/pyauth/exile/coverage.svg?branch=master
        :target: https://codecov.io/github/pyauth/exile?branch=master
.. image:: https://img.shields.io/pypi/v/exile.svg
        :target: https://pypi.python.org/pypi/exile
.. image:: https://img.shields.io/pypi/l/exile.svg
        :target: https://pypi.python.org/pypi/exile
.. image:: https://readthedocs.org/projects/exile/badge/?version=latest
        :target: https://exile.readthedocs.io/


Changes.rst
Changes for v0.1.1 (2021-09-24)
===============================

-  Properly recognize newer yubikeys

-  Add support for SET_CODE and VALIDATE

Changes for v0.1.0 (2019-03-03)
===============================

-  Add TOTP helpers and docs

-  Manage multiple yubikeys and absent key condition

-  Include signers in package

-  Support HmacV1 and multipart responses

-  Add Linux install instructions

Changes for v0.0.3 (2019-02-26)
===============================

-  Begin tests

Changes for v0.0.2 (2019-02-26)
===============================

-  Initial release.



test/test.py
#!/usr/bin/env python

import os, sys, unittest, json, collections, base64, datetime
import boto3, botocore.auth

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))  # noqa

from exile import YKOATH, TOTP, SCardManager, botocore_signers

class TestExile(unittest.TestCase):
    def test_scard_manager(self):
        for reader in SCardManager():
            with reader:
                pass

    def test_exile_totp(self):
        TOTP().save("google", "JBSWY3DPEHPK3PXP")
        TOTP().get("google")
        TOTP().verify("260153", label="google", at=datetime.datetime.fromtimestamp(1297553958))
        otpauth_uri = 'otpauth://totp/Secure%20App:alice%40google.com?secret=JBSWY3DPEHPK3PXP&issuer=Secure%20App'
        TOTP().save_otpauth_uri(otpauth_uri)
        TOTP().verify("260153", label="Secure%20App:alice%40google.com", at=datetime.datetime.fromtimestamp(1297553958))

    def write_active_aws_key_to_yubikey(self):
        credentials = boto3.Session().get_credentials()

        key_name = "exile-{}-SigV4".format(credentials.access_key)
        secret = b"AWS4" + credentials.secret_key.encode()
        YKOATH().put(key_name, secret, algorithm=YKOATH.Algorithm.SHA256)

        key_name = "exile-{}-HmacV1".format(credentials.access_key)
        secret = credentials.secret_key.encode()
        YKOATH().put(key_name, secret, algorithm=YKOATH.Algorithm.SHA1)

    def test_exile(self):
        self.write_active_aws_key_to_yubikey()
        botocore_signers.install()

        boto3.client("sts").get_caller_identity()
        boto3.client("s3").generate_presigned_url(ClientMethod="get_object", Params={"Bucket": "foo", "Key": "bar"})

if __name__ == '__main__':
    unittest.main()


exile/ykoath/const.py
from enum import Enum

class YKOATHConstants:
    device_prefix = "yubico yubikey"
    HMAC_MINIMUM_KEY_SIZE = 14

    class Tag:
        NAME = 0x71
        NAME_LIST = 0x72
        KEY = 0x73
        CHALLENGE = 0x74
        RESPONSE = 0x75
        TRUNCATED_RESPONSE = 0x76
        NO_RESPONSE = 0x77
        PROPERTY = 0x78
        VERSION = 0x79
        IMF = 0x7a
        ALGORITHM = 0x7b
        TOUCH = 0x7c

    class OATHType(Enum):
        HOTP = 0x10
        TOTP = 0x20

    class Properties:
        REQUIRE_TOUCH = 0x02

    class Instruction:
        SELECT = 0xa4
        PUT = 0x01
        DELETE = 0x02
        SET_CODE = 0x03
        RESET = 0x04
        LIST = 0xa1
        CALCULATE = 0xa2
        VALIDATE = 0xa3
        CALCULATE_ALL = 0xa4
        SEND_REMAINING = 0xa5

    class Algorithm(Enum):
        SHA1 = 0x01
        SHA256 = 0x02
        SHA512 = 0x03

    class Application:
        OTP = b'\xa0\x00\x00\x05\x27\x20\x01'
        MGR = b'\xa0\x00\x00\x05\x27\x47\x11\x17'
        OPGP = b'\xd2\x76\x00\x01\x24\x01'
        OATH = b'\xa0\x00\x00\x05\x27\x21\x01'
        PIV = b'\xa0\x00\x00\x03\x08'
        U2F = b'\xa0\x00\x00\x06\x47\x2f\x00\x01'

    class Response(Enum):
        SUCCESS = b'\x90\x00'
        NO_SPACE = b'\x6a\x84'
        NOT_FOUND = b'\x69\x84'
        AUTH_REQUIRED = b'\x69\x82'
        WRONG_SYNTAX = b'\x6a\x80'
        GENERIC_ERROR = b'\x65\x81'
        MORE_DATA_AVAILABLE = b'\x61'


exile/botocore_signers.py
import botocore.auth
from botocore.compat import encodebytes
from . import YKOATH

class YKSigV4Auth(botocore.auth.SigV4Auth):
    def signature(self, string_to_sign, request):
        if not hasattr(self, "_ykoath"):
            self._ykoath = YKOATH()
        key_name = "exile-{}-SigV4".format(self.credentials.access_key)
        k_date = self._ykoath.calculate(key_name,
                                        request.context["timestamp"][0:8].encode(),
                                        want_truncated_response=False)
        k_region = self._sign(k_date, self._region_name)
        k_service = self._sign(k_region, self._service_name)
        k_signing = self._sign(k_service, "aws4_request")
        return self._sign(k_signing, string_to_sign, hex=True)

class YKHmacV1Auth(botocore.auth.HmacV1Auth):
    def sign_string(self, string_to_sign):
        if not hasattr(self, "_ykoath"):
            self._ykoath = YKOATH()
        key_name = "exile-{}-HmacV1".format(self.credentials.access_key)
        digest = self._ykoath.calculate(key_name, string_to_sign.encode(), want_truncated_response=False)
        return encodebytes(digest).strip().decode("utf-8")

def install():
    botocore.auth.SigV4Auth.signature = YKSigV4Auth.signature
    botocore.auth.HmacV1Auth.sign_string = YKHmacV1Auth.sign_string


exile/ykoath/__init__.py
import base64, struct, typing, hashlib, hmac
from collections import namedtuple
from datetime import datetime
from urllib.parse import urlparse, parse_qs
from ..exceptions import YKOATHError
from ..scard import i2b, SCardManager, SCardReader
from .const import YKOATHConstants

YKOATHCredential = namedtuple("YKOATHCredential", ("name", "oath_type", "algorithm"))

class YKOATH(YKOATHConstants):
    """
    See https://developers.yubico.com/OATH/YKOATH_Protocol.html
    """
    def __init__(self, device: SCardReader = None, password: str = None) -> None:
        if device is None:
            for reader in SCardManager():
                if reader.name.lower().startswith(self.device_prefix):
                    device = reader
                    break
            else:
                raise YKOATHError("No YubiKey found")
        self.device = device
        res = self.send_apdu(cla=0, ins=self.Instruction.SELECT, p1=0x04, p2=0, data=self.Application.OATH)
        _, self._version, res = self.parse_tlv(res, self.Tag.VERSION)
        _, self._id, res = self.parse_tlv(res, self.Tag.NAME)
        _, self._challenge, res = self.parse_tlv(res)
        if self._challenge and password is not None:
            self.validate(password)

    def send_apdu(self, **kwargs):
        with self.device:
            res = self.device.send_apdu(**kwargs)
            while res[-2:-1] == self.Response.MORE_DATA_AVAILABLE.value:
                res = res[:-2] + self.device.send_apdu(cla=0, ins=self.Instruction.SEND_REMAINING, p1=0, p2=0, data=b"")
        if res[-2:] != self.Response.SUCCESS.value:
            raise YKOATHError(self.Response(res[-2:]))
        return res

    def parse_tlv(self, data, expect_tag=None):
        assert isinstance(data, bytes)
        tag, length = data[0], data[1]
        if expect_tag:
            assert tag == expect_tag
        value, data = data[2:2 + length], data[2 + length:]
        return tag, value, data

    def put(self, credential_name: str, secret: bytes, require_touch=False,
            oath_type=YKOATHConstants.OATHType.TOTP, algorithm=YKOATHConstants.Algorithm.SHA1, digits=6):
        secret_header = i2b(oath_type.value | algorithm.value) + i2b(digits)
        # secret = hmac_shorten_key(secret, algorithm)
        secret = secret.ljust(self.HMAC_MINIMUM_KEY_SIZE, b'\x00')
        data = i2b(self.Tag.NAME) + i2b(len(credential_name)) + credential_name.encode()
        data += i2b(self.Tag.KEY) + i2b(len(secret_header) + len(secret)) + secret_header + secret
        if require_touch:
            data += i2b(self.Tag.PROPERTY) + i2b(self.Properties.REQUIRE_TOUCH)
        return self.send_apdu(cla=0, ins=self.Instruction.PUT, p1=0, p2=0, data=data)

    def delete(self, credential_name: str):
        data = i2b(self.Tag.NAME) + i2b(len(credential_name)) + credential_name.encode()
        return self.send_apdu(cla=0, ins=self.Instruction.DELETE, p1=0, p2=0, data=data)

    def reset(self):
        return self.send_apdu(cla=0, ins=self.Instruction.RESET, p1=0xde, p2=0xad, data=b"")

    def list(self):
        return self.send_apdu(cla=0, ins=self.Instruction.LIST, p1=0, p2=0, data=b"")

    def calculate(self, credential_name: str, challenge: typing.Union[bytes, int], want_truncated_response=True):
        chal_bytes = challenge if isinstance(challenge, bytes) else int_to_bytestring(challenge)
        data = i2b(self.Tag.NAME) + i2b(len(credential_name)) + credential_name.encode()
        data += i2b(self.Tag.CHALLENGE) + i2b(len(chal_bytes)) + chal_bytes
        p2 = 0x01 if want_truncated_response else 0
        res = self.send_apdu(cla=0, ins=self.Instruction.CALCULATE, p1=0, p2=p2, data=data)
        assert res[0] == self.Tag.TRUNCATED_RESPONSE if want_truncated_response else self.Tag.RESPONSE
        res_len, digits = res[1], res[2]
        if want_truncated_response:
            return str(struct.unpack('>I', res[3:3 + res_len - 1])[0]).zfill(digits)
        else:
            return res[3:3 + res_len - 1]

    def set_code(self, password):
        key = hashlib.pbkdf2_hmac('sha256', password.encode(), self._id, 1000)
        test_challenge = b'01234567'
        test_response = hmac.new(key, test_challenge, 'sha256').digest()
        data = i2b(self.Tag.KEY) + i2b(len(key) + 1) + i2b(self.Algorithm.SHA256.value) + key
        data += i2b(self.Tag.CHALLENGE) + i2b(len(test_challenge)) + test_challenge
        data += i2b(self.Tag.RESPONSE) + i2b(len(test_response)) + test_response
        return self.send_apdu(cla=0, ins=self.Instruction.SET_CODE, p1=0, p2=0, data=data)

    def validate(self, password):
        key = hashlib.pbkdf2_hmac('sha256', password.encode(), self._id, 1000)
        response = hmac.new(key, self._challenge, 'sha256').digest()
        data = i2b(self.Tag.RESPONSE) + i2b(len(response)) + response
        data += i2b(self.Tag.CHALLENGE) + i2b(len(self._challenge)) + self._challenge
        return self.send_apdu(cla=0, ins=self.Instruction.VALIDATE, p1=0, p2=0, data=data)

    def __iter__(self):
        res = self.list()[:-2]
        while res:
            name_tag, name_len, alg = res[0], res[1], res[2]
            assert name_tag == self.Tag.NAME_LIST
            algorithm, oath_type = self.Algorithm(alg & 0x0f), self.OATHType(alg & 0xf0)
            name_data = res[3:3 + name_len - 1]
            yield YKOATHCredential(name=name_data.decode(), oath_type=oath_type, algorithm=algorithm)
            res = res[name_len + 2:]

def int_to_bytestring(i: int, padding=8):
    result = bytearray()
    while i != 0:
        result.append(i & 0xFF)
        i >>= 8
    return bytes(bytearray(reversed(result)).rjust(padding, b'\0'))

class TOTP(YKOATH):
    default_time_step = 30

    def save(self, label: str, secret: str):
        self.put(label, base64.b32decode(secret, casefold=True))

    def save_otpauth_uri(self, otpauth_uri: str):
        otpauth = urlparse(otpauth_uri)
        assert otpauth.scheme == "otpauth"
        assert otpauth.netloc == "totp"
        label = otpauth.path.lstrip("/")
        secret = parse_qs(otpauth.query)["secret"][0]
        self.save(secret=secret, label=label)

    def get(self, label: str, at: datetime = None, time_step: int = default_time_step):
        if at is None:
            at = datetime.now()
        return self.calculate(label, int(at.timestamp() / time_step))

    def verify(self, code: str, label: str, at: datetime = None, time_step: int = default_time_step):
        if self.get(label=label, at=at, time_step=time_step) != code:
            raise YKOATHError("TOTP code mismatch")
